src/components/checkboxTree/utils.ts

```diff
@@ -0,0 +1,117 @@
+import { Composition } from "./types";
+
+export const preProcessTree = (
+  tree: Composition[],
+  preferredSeperator: string
+): Composition[] => {
+  const preservedTree = JSON.parse(JSON.stringify(tree));
+  const mappedUids = getRelations(tree, preferredSeperator);
+
+  return attachRelations(preservedTree, mappedUids, preferredSeperator);
+};
+
+export const getChildrenUids = (composition: Composition): string[] => {
+  const uids: string[] = [];
+  if (composition.children) {
+    const recursive = (children: Composition[]) => {
+      children.forEach((child) => {
+        uids.push(child.uid);
+        if (child.children) {
+          recursive(child.children);
+        }
+      });
+    };
+    recursive(composition.children);
+    return uids;
+  }
+
+  return [];
+};
+
+export const getParentsUids = (
+  compositionUid: string,
+  preferredSeperator: string
+): string[] => {
+  const currentIdSpilt = compositionUid.split(preferredSeperator);
+
+  return currentIdSpilt.map((_, index) =>
+    [...currentIdSpilt].splice(0, index + 1).join(preferredSeperator)
+  );
+};
+
+export const getRelations = (
+  tree: Composition[],
+  preferredSeperator = "->"
+) => {
+  const allRelations = [];
+
+  let branch: string[] = [];
+  let treeCopy = [...tree];
+  let isTreeEmpty = false;
+
+  const firstChildOfItem = (childs: Composition[]) => {
+    const firstChild = childs ? childs[0] : null;
+    if (!firstChild) {
+      return;
+    }
+    branch.push(firstChild.uid);
+    if (firstChild.children && firstChild.children.length > 0) {
+      firstChildOfItem(firstChild.children);
+    } else {
+      treeCopy = removeObject(treeCopy, firstChild.uid);
+    }
+  };
+
+  while (!isTreeEmpty) {
+    firstChildOfItem(treeCopy);
+    allRelations.push([...new Set(branch)].join(preferredSeperator));
+    branch = [];
+    isTreeEmpty = treeCopy.length === 0;
+  }
+
+  return allRelations;
+};
+
+const removeObject = (tree: Composition[], uid: string) => {
+  const treeCopy = [...tree];
```

Comment:
I think spread operator does a shallow copy, is that fine in this case?

You could use lodash `cloneDeep`.

Reply:
Yep, this object should shrink with each iteration, but I've noticed that there is no need to copy it at all. So removed the setting of a copy.